Add a measure tool to elevation profile widget

nyalldawson opened the pull request:
Allows measuring distances on the plot

nyalldawson committed:
[feature] Add a measure tool to elevation profile widget

Allows measuring distances on the plot

uclaros commented on src/app/elevation/qgselevationprofiletoolmeasure.cpp:
```diff
@@ -0,0 +1,125 @@
+  const QPointF snappedPoint = event->snappedPoint().toQPointF();
+  const QgsProfilePoint endPoint = mElevationCanvas->canvasPointToPlotPoint( constrainPointToRect( snappedPoint, plotArea ) );
+
+  mRubberBand->setLine( QLineF( mRubberBand->line().p1(), mElevationCanvas->mapToScene( snappedPoint.toPoint() ) ) );
```
`mRubberBand->line().p1()` is the plot point. It needs to be recalculated from `mStartPoint` since the canvas may have zoomed/panned in the meantime.

nyalldawson replied:
Fixed

uclaros commented on src/app/elevation/qgselevationprofilewidget.cpp:
```diff
@@ -237,6 +250,22 @@ QgsElevationProfileWidget::QgsElevationProfileWidget( const QString &name )
   topLayout->addStretch( 1 );
+
+  mInfoLabel = new QLabel();
+  topLayout->addWidget( mInfoLabel );
```
Is this the best place to show display the result?
I think many users would benefit from also separately displaying `dZ` and `dXY` values.
Actually `dZ` is the most important value here, since other distances can me measured on the main canvas!

nyalldawson replied:
> Is this the best place to show display the result?

It's not - but at this stage I'm so over budget for this project that I can't afford to tweak further

wonder-sk replied:
Agreed with Stefanos that it is very desirable to have elevation and 2d distance reported also separately (e.g. if I need to find out building/tree height, without having to worry whether my measure line is perfectly vertical)

nyalldawson replied:
Ok, i've gone for the same approach as the canvas measure tool uses now

uclaros reviewed (commented):
Sweet!
Some usability tweaks are needed imho though.

nyalldawson commented on src/app/elevation/qgselevationprofiletoolmeasure.cpp:
```diff
@@ -0,0 +1,125 @@
+  emit distanceChanged( distance );
+}
+
+void QgsElevationProfileToolMeasure::plotPressEvent( QgsPlotMouseEvent *event )
```
My opinion is that this is an error in our other tools -- if you precisely position the cursor then press down, it's more likely that the press location is accurate and that little unwanted movements will occur before release. It's my understanding that most applications use press coordinates for this reason. So to me this is a bug in all the other tools. 😱

uclaros replied:
We kind of started a discussion for that on #37119 some time ago :)
I believe homogeneity is key to good ux, even in the small stuff, that's why I suggested the change.

nyalldawson committed:
Add signal when plot area is changed

nyalldawson committed:
Remove duplicate code

nyalldawson committed:
Make private method public

nyalldawson committed:
Fix plot area calculation

nyalldawson committed:
Make QgsClipper::clipLineSegment public

nyalldawson committed:
Ensure measure line stays in the correct location when plot is
panned/zoomed

nyalldawson committed:
Immediately reset distance on start

nyalldawson committed:
Clear measure line on right click

nyalldawson commented on src/app/elevation/qgselevationprofiletoolmeasure.cpp:
```diff
@@ -0,0 +1,125 @@
+
+void QgsElevationProfileToolMeasure::plotPressEvent( QgsPlotMouseEvent *event )
+{
+  if ( event->button() != Qt::LeftButton )
+  {
+    event->ignore();
+    return;
+  }
```
Thanks, done

nyalldawson committed:
Use a dialog to show measure results, and include elevation and curve
distance delta measures

nyalldawson commented:
Here's how things look now

https://user-images.githubusercontent.com/1829991/167053108-1e1a594d-b340-4150-a18a-19232c9e13a9.mp4

nyalldawson committed:
Show curve distance measure using unit types

We CAN'T show units for the total distance or elevation measure, as
these measurements are currently inherently dimensionless. We need
proper support for vertical CRS before we can support that!

nyalldawson committed:
Localisation

nyalldawson committed:
More test tolerance

uclaros reviewed (approved):
<3 @nyalldawson !
Just a failing test pending

nyalldawson committed:
More tolerance

nyalldawson committed:
Test tolerance

nyalldawson committed:
Tolerance

nyalldawson committed:
Tolerance
